crates/cubecl-runtime/src/tune/tuner.rs

```diff
@@ -47,115 +52,146 @@ impl<K: AutotuneKey> Tuner<K> {
+                .await
+                .expect("Autotune results channel closed");
+        };
 
-        match self.tune_cache.try_cache(autotune_operation_set) {
-            super::TuneCacheResult::Hit(ops) => ops,
-            super::TuneCacheResult::Miss(_) => panic!("We just inserted, should not miss"),
-        }
+        // Spawn the tuning as a task.
+        #[cfg(target_family = "wasm")]
+        wasm_bindgen_futures::spawn_local(fut);
+
+        // It is totally possible here to run the tuning on a thread, which woulp startup time,
+        // but might have two downsides:
+        // - Benchmarks now need a "warmup" time until a good kernel is selected.
+        // - Tuning might be less precise, as it's possible that other operations are
+        //   submitted while tuning, which might skew results.
+        #[cfg(not(target_family = "wasm"))]
+        futures_lite::future::block_on(fut);
     }
 
-    fn run_benchmark<S, C, Out>(
-        &mut self,
+    async fn run_benchmark<S: ComputeServer, C: ComputeChannel<S>, Out>(
         operation: Box<dyn AutotuneOperation<Out>>,
         client: &ComputeClient<S, C>,
-    ) -> Result<BenchmarkDurations, BenchError>
-    where
-        S: ComputeServer,
-        C: ComputeChannel<S>,
-    {
-        #[cfg(feature = "std")]
+    ) -> Result<BenchmarkDurations, BenchError> {
+        let tuner = TuneBenchmark::new(operation, client.clone());
+
+        #[cfg(any(target_family = "wasm", not(feature = "std")))]
         {
-            catch_unwind(AssertUnwindSafe(|| {
-                TuneBenchmark::new(operation, client.clone()).run()
-            }))
-            .map_err(|e| {
-                println!("Caught error while benchmarking, falling back to next operation.");
-                ManuallyDrop::new(e)
-            })
+            Ok(tuner.sample_durations().await)
         }
-        #[cfg(not(feature = "std"))]
-        Ok(TuneBenchmark::new(operation, client.clone()).run())
-    }
-
-    fn find_fastest(&self, results: Vec<Result<BenchmarkDurations, BenchError>>) -> usize {
-        let mut smallest_duration = Duration::MAX;
-        let mut fastest_tunable = None;
-
-        for (i, result) in results.into_iter().enumerate() {
-            let result = match result {
-                Ok(result) => result,
-                Err(_) => continue,
-            };
-            let computed = BenchmarkComputations::new(&result);
 
-            if computed.median < smallest_duration {
-                smallest_duration = computed.median;
-                fastest_tunable = Some(i);
-            }
+        #[cfg(all(not(target_family = "wasm"), feature = "std"))]
+        {
+            use futures_lite::FutureExt;
+            AssertUnwindSafe(tuner.sample_durations())
+                .catch_unwind()
+                .await
+                .map_err(|e| {
+                    println!("Caught error while benchmarking, falling back to next operation.");
```

Comment:
It should be a log instead.

Reply:
Ah yes - made it a log::warn because panic unwinds aren't supported on wasm/no-std so shouldn't rely on them too much.